Fix for multibag replay stagnation

MichaelOrlov opened the pull request:
## Description

This PR fixes a problem for slow playback when one reader has a lower frequency topic than others

- Added cache for the last element from each reader and taking messages from cache in a chronological order instead of the round-robin order.
- To mitigate "spaggetish" code after the fix, this PR also introduces the `ReadersWrapper` class, which will encapsulate readers with their options and message cache.
- The `ReadersWrapper` class manages multiple `rosbag2_cpp::Reader` instances, allowing for chronological reading of messages across all readers. It maintains a cache of the next message from each reader and provides methods to retrieve the next message in chronological order, seek to a specific timestamp across all readers, and apply filters to all readers. It also provides access to the earliest and latest timestamps across all readers.

- Fixes #2156
- Depends on the #2157 

### Is this user-facing behavior change?
Fixed a bug where playback of the higher frequency topics can slow down when present slow frequency topics in another playing bag.

### Did you use Generative AI?
Yes. I am using GitHub Copilot, GPT-4.1, in my workflow. in particular to help with unit tests and documentation.

### Additional Information
This PR can be backported to the other distros.

MichaelOrlov committed:
Add "high_freq_topics_does_not_starve_in_multibag_playback" test

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Quick fix for slow playback when one reader has lower freq topic

- Add cache for the last element from each reader and taking messages
 from cache in a chronological order.
- Also use "rcpputils::unique_lock" consistently for the reader_mutex_

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Add ReadersWrapper class that encapsulates readers and messages cache

- The ReadersWrapper class manages multiple rosbag2_cpp::Reader
 instances, allowing for chronological reading of messages across all
 readers. It maintains a cache of the next message from each reader and
 provides methods to retrieve the next message in chronological order,
 seek to a specific timestamp across all readers, and apply filters to
 all readers. It also provides access to the earliest and latest
 timestamps across all readers.

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Add unit tests for ReadersWrapper class

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Delete default ctor and copy/move operators for ReadersWrapper class

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov commented:
@ros-pull-request-builder retest this please

MichaelOrlov commented:
Pulls: ros2/rosbag2#2158
Gist: https://gist.githubusercontent.com/MichaelOrlov/15d3a9082ed76aa64911d78bc47d74f7/raw/bc34f98f1ab92956f6be1d973f053311a2f6ec79/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_transport rosbag2_tests
TEST args:  --packages-above rosbag2_transport rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16834
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24744)](http://ci.ros2.org/job/ci_linux/24744/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18773)](http://ci.ros2.org/job/ci_linux-aarch64/18773/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4194)](http://ci.ros2.org/job/ci_linux-rhel/4194/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25260)](http://ci.ros2.org/job/ci_windows/25260/)

fujitatomoya commented on rosbag2_transport/include/rosbag2_transport/readers_wrapper.hpp:
```diff
@@ -0,0 +1,131 @@
+/// retrieve the next message in chronological order, seek to a specific timestamp across all
+/// readers, and apply filters to all readers. It also provides access to the earliest and
+/// latest timestamps across all readers.
+class ROSBAG2_TRANSPORT_PUBLIC ReadersWrapper
```
Just a minor suggestion. How about naming the class `ReaderManager` would provide clearer intent, as the class is responsible for managing multiple `rosbag2_cpp::Reader` instances such as opening, seeking, filtering, and reading messages chronologically. IMO "Wrapper" sounds more general pattern and doesn’t convey what the class actually does, while "Manager" directly indicates its role in handling and coordinating the readers?

MichaelOrlov replied:
@fujitatomoya Thank you for your review. Your suggestion about renaming to the `ReadersManager` does make sense to me. I will go ahead and rename it.

fujitatomoya reviewed (approved):
lgtm with green CI.

MichaelOrlov committed:
Rename ReadersWrapper class to the ReadersManager

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov committed:
Rename get_next_chronological_message_from_cache

- Reasoning: Not mentioning cache in the name because this is some
 implementation details and could be changed in the future. However,
 the API needs to be stable and clear.
- Also added a comment about reasoning for usage pre-cache inside
 implementation file.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
Re-run CI after renames
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16837
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24747)](http://ci.ros2.org/job/ci_linux/24747/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18776)](http://ci.ros2.org/job/ci_linux-aarch64/18776/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4198)](http://ci.ros2.org/job/ci_linux-rhel/4198/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25264)](http://ci.ros2.org/job/ci_windows/25264/)

MichaelOrlov commented:
Hold off from merging this PR. We have regression in the tests under the load
```bash
/tmp/ws/src/rosbag2/rosbag2_transport/test/rosbag2_transport/test_play_seek.cpp:144: Failure
Expected equality of these values:
  replayed_topic1[2]->int32_value
    Which is: 4
  1

/tmp/ws/src/rosbag2/rosbag2_transport/test/rosbag2_transport/test_play_seek.cpp:145: Failure
Expected equality of these values:
  replayed_topic1[3]->int32_value
    Which is: 1
  2
```

### Preliminary analysis
That is because a data race for the internal queue cleanup and refill.
Before, the queue cleanup and refill was protected by the `readers_mutex`; now I moved those mutex inside `ReadersManager`, and when we are doing `seek()` we need to do `queue->purge()` and readers->seek(timestamp). However, another thread still trying to refill the queue from the readers in the background.

MichaelOrlov committed:
Rename "no_messages_in_cache" to the "has_next()"

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov committed:
Renames for mutex and condition_variable

- Rename "skip_message_in_main_play_loop_mutex_" to the
 "main_play_loop_mutex_" and "playback_finished_cv_" into the
 "is_in_playback_cv_".

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov committed:
Fix for messages_queue access race in seek(time)

- Stop storage loading thread before doing seek to avoid race condition
- Return from seek at the beginning if player is not in playback mode.
  Reasoning: We will seek to the beginning of the bag when will start
  play() anyway.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov committed:
Fix for play_fails_gracefully_if_needed_coverter_plugin_does_not_exist

Add missing converter options when open readers

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov committed:
Speedup "load_storage_content()" interruption

- Check "load_storage_content_" inside "enqueue_up_to_boundary()"

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
Note: Decided to move a couple of commits 
-  [Fix multiple race conditions in the play_messages_from_queue()](https://github.com/ros2/rosbag2/pull/2158/commits/48fa0bc8c1877ba8e3e966757fd6469f06d17aa0)
- [Wait for Player->play() to be called in the wait_for_playback_to_start()](https://github.com/ros2/rosbag2/pull/2158/commits/7ec6449a75373765133c01d18ffa4bd02f3f736e)
to the follow-up PR, since they are unrelated to the original fix and could be merged separately.

MichaelOrlov commented:
Re-run CI
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16885
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24800)](http://ci.ros2.org/job/ci_linux/24800/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18829)](http://ci.ros2.org/job/ci_linux-aarch64/18829/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4250)](http://ci.ros2.org/job/ci_linux-rhel/4250/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25312)](http://ci.ros2.org/job/ci_windows/25312/)

MichaelOrlov commented:
The Linux-RHEL CI job failed due to not found Uncrustify.
Trying to re-run this job
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4259)](http://ci.ros2.org/job/ci_linux-rhel/4259/)

MichaelOrlov commented:
@mjcarroll @ahcorde Any ideas for why [RHEL CI job](https://ci.ros2.org/job/ci_linux-rhel/4259/#showFailuresLink) keep fails?  
Something is wrong with uncrustify.  
I am curious if it is somehow related to the recent removal of the `uncrustify_vendor` package.

MichaelOrlov commented:
Pulls: ros2/rosbag2#2158
Gist: https://gist.githubusercontent.com/MichaelOrlov/2ba68fbf0fbdc4a9da5643bfca61fc85/raw/bc34f98f1ab92956f6be1d973f053311a2f6ec79/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_transport rosbag2_tests
TEST args:  --packages-above rosbag2_transport rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16939
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24858)](http://ci.ros2.org/job/ci_linux/24858/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18891)](http://ci.ros2.org/job/ci_linux-aarch64/18891/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4313)](http://ci.ros2.org/job/ci_linux-rhel/4313/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25369)](http://ci.ros2.org/job/ci_windows/25369/)

MichaelOrlov commented:
The RHEL CI job failed due to the CI infrastructure issue. It was auto-restarted and passed green
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=4314)](http://ci.ros2.org/job/ci_linux-rhel/4314/)

MichaelOrlov commented:
The CI is green after reverting the `uncrustify_vendor` removal. Merging then.

MichaelOrlov commented:
https://github.com/Mergifyio backport kilted jazzy
